### ISSUE TITLE
Mixer may report incorrect referenced attribute

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#6552]
Proxy log:
```
[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:150] Response referenced mismatchs with request
[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:151] Request attributes: attributes {
  key: "connection.mtls"
  value {
    bool_value: false
  }
}
attributes {
  key: "context.protocol"
  value {
    string_value: "http"
  }
}
attributes {
  key: "context.reporter.local"
  value {
    bool_value: false
  }
}
attributes {
  key: "context.reporter.uid"
  value {
    string_value: "kubernetes://istio-ingressgateway-ff8dbcd7c-hmvwt.istio-system"
  }
}
attributes {
  key: "destination.service"
  value {
    string_value: "productpage.istio-system.svc.cluster.local"
  }
}
attributes {
  key: "destination.service.host"
  value {
    string_value: "productpage.istio-system.svc.cluster.local"
  }
}
attributes {
  key: "destination.service.name"
  value {
    string_value: "productpage"
  }
}
attributes {
  key: "destination.service.namespace"
  value {
    string_value: "istio-system"
  }
}
attributes {
  key: "destination.service.uid"
  value {
    string_value: "istio://istio-system/services/productpage"
  }
}
attributes {
  key: "request.headers"
  value {
    string_map_value {
      entries {
        key: ":authority"
        value: "35.233.147.55"
      }
      entries {
        key: ":method"
        value: "GET"
      }
      entries {
        key: ":path"
        value: "/productpage"
      }
      entries {
        key: "accept-encoding"
        value: "gzip"
      }
      entries {
        key: "user-agent"
        value: "Go-http-client/1.1"
      }
      entries {
        key: "x-b3-sampled"
        value: "1"
      }
      entries {
        key: "x-b3-spanid"
        value: "85c2bdb533db84a7"
      }
      entries {
        key: "x-b3-traceid"
        value: "85c2bdb533db84a7"
      }
      entries {
        key: "x-envoy-decorator-operation"
        value: "productpage.istio-system.svc.cluster.local:9080/productpage"
      }
      entries {
        key: "x-envoy-internal"
        value: "true"
      }
      entries {
        key: "x-forwarded-for"
        value: "10.138.0.3"
      }
      entries {
        key: "x-forwarded-proto"
        value: "http"
      }
      entries {
        key: "x-request-id"
        value: "2f785901-1e25-9647-99b5-b2b6c9fffee9"
      }
    }
  }
}
attributes {
  key: "request.host"
  value {
    string_value: "35.233.147.55"
  }
}
attributes {
  key: "request.method"
  value {
    string_value: "GET"
  }
}
attributes {
  key: "request.path"
  value {
    string_value: "/productpage"
  }
}
attributes {
  key: "request.scheme"
  value {
    string_value: "http"
  }
}
attributes {
  key: "request.time"
  value {
    timestamp_value {
      seconds: 1529604800
      nanos: 818476127
    }
  }
}
attributes {
  key: "request.useragent"
  value {
    string_value: "Go-http-client/1.1"
  }
}
attributes {
  key: "source.uid"
  value {
    string_value: "kubernetes://istio-ingressgateway-ff8dbcd7c-hmvwt.istio-system"
  }
}

[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:152] Referenced attributes: Absence-keys: adapter_template_kubernetes.output.destination_container_name, adapter_template_kubernetes.output.destination_labels, adapter_template_kubernetes.output.destination_namespace, adapter_template_kubernetes.output.destination_owner, adapter_template_kubernetes.output.destination_pod_ip, adapter_template_kubernetes.output.destination_pod_name, adapter_template_kubernetes.output.destination_service_account_name, adapter_template_kubernetes.output.destination_workload_name, adapter_template_kubernetes.output.destination_workload_namespace, adapter_template_kubernetes.output.destination_workload_uid, destination.port, destination.uid, source.ip, Exact-keys: context.protocol, destination.service, request.headers[x-user], source.uid, 
```

Note that that `request.headers[x-user]` is marked as exact incorrectly. 

Log: https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6338/e2e-mixer-no_auth/113/

**Comment by kyessenov** [comment]
Seems related to internal mixer caching, but I can't quite isolate the code responsible for this.

**Comment by kyessenov** [comment]
cc @douglas-reid @geeknoid 
I've spent a few hours today trying to reproduce in a unit test, or find an obvious error in the relevant code, but no luck.

**Comment by geeknoid** [comment]
I think it's driven by the specific way the config references the attribute.

request.headers["x-user"] != ""
request.headers["x-user" != "ABC"
request.headers["x-user"] == ""
request.headers["x-user" == "ABC"

**Comment by kyessenov** [comment]
Evaluation seems to be not conditional on the binary operation. There's going to be symbol resolution for the map, in all of the cases. Is there some optimizer running that can erase sub-expressions?

**Comment by kyessenov** [comment]
I don't believe this has been solved.

**Comment by jsw** [comment]
I think I am seeing this in 1.0.4.

```
Sample errors:
[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:152] Referenced attributes: Absence-keys: source.ip, Exact-keys: context.protocol, context.reporter.kind, destination.namespace, destination.port, destination.service, destination.uid, source.uid,
[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:150] Response referenced mismatchs with request
[libprotobuf ERROR src/istio/mixerclient/check_cache.cc:151] Request attributes: attributes {
```
Let me know if I can provide more info.

**Comment by kyessenov** [comment]
possibly related to #9596 ?
TL;DR mixer server side cache ignores string maps in the cache key computation. The workaround is to disable it with --numCheckCacheEntries=0 in mixer deployment.

**Comment by kyessenov** [comment]
Closing in favor of #9596
